### PR TITLE
`<format>`: crash when formatting a floating-point value with #/L and a large precision

### LINKED ISSUE #4320: <format>: crash when formatting a floating-point value with `#`/`L` and a large precision
# Describe the bug
`<format>` uses a loop to find the beginning of the exponent part in order to insert trailing zeroes. But when `_Specs._Alt || _Specs._Localized` is true, the position has already been stored in `_Exponent_start`. The loop actually moves `_Exponent_start` away from the correct position, and eventually causes a crash due to out-of-bound access.

https://github.com/microsoft/STL/blob/3eac329d1f614ecf138d96c22a3b02f87076bc4a/stl/inc/format#L3108-L3112

https://github.com/microsoft/STL/blob/3eac329d1f614ecf138d96c22a3b02f87076bc4a/stl/inc/format#L3087-L3089

# Command-line test case
```
D:\test>type test-format-3.cpp
#include <format>
#include <iostream>

int main() {
    std::cout << std::format("{:#.1075e}\n", 1.0);
}

D:\test>cl /std:c++20 /EHs /utf-8 test-format-3.cpp
Microsoft (R) C/C++ Optimizing Compiler Version 19.39.33321 for x64
Copyright (C) Microsoft Corporation.  All rights reserved.

test-format-3.cpp
Microsoft (R) Incremental Linker Version 14.39.33321.0
Copyright (C) Microsoft Corporation.  All rights reserved.

/out:test-format-3.exe
test-format-3.obj

D:\test>test-format-3

D:\test>echo %errorlevel%
-1073741819
```

# Expected behavior
no crash

# STL version

    https://github.com/microsoft/STL/commit/3eac329d1f614ecf138d96c22a3b02f87076bc4a

### PR DESCRIPTION
**Pull request by Andor233**
Fix #4320

### CODE REVIEW

File: tests/std/tests/P0645R10_text_formatting_formatting/test.cpp [hunk 1]
```diff
@@ -1520,6 +1520,23 @@ void test_gh_4319() {
     assert(format(STR("{:}"), 12345678.0) == STR("12345678"));
 }
 
+// GH-4320: <format>: crash when formatting a floating-point value with #/L and a large precision
+void test_gh_4320() {
```
Comment (frederick-vs-ja) [inline tests/std/tests/P0645R10_text_formatting_formatting/test.cpp:1524]: Should we also test the `wchar_t` version?

Comment (cpplearner) [reply tests/std/tests/P0645R10_text_formatting_formatting/test.cpp:1524]: IMO it's unnecessary. We know that the affected code path does not depend on the character type.

**Review by cpplearner — APPROVED** [review]

File: tests/std/tests/P0645R10_text_formatting_formatting/test.cpp (hunk 1 above)
Comment (Andor233) [reply tests/std/tests/P0645R10_text_formatting_formatting/test.cpp:1525]: But I still add it, `we know` doesn't means still same in the future.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Perfect, thank you! :heart_eyes_cat:

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by Andor233** [comment]
> I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

Still work on Sunday?

**Comment by StephanTLavavej** [comment]
I am a notoriously bad kitty, jumping up on tables where I'm not allowed, and working on the weekends when I should be relaxing :smirk_cat:

Thanks for fixing this crash in one of C++20/23's most popular components! :hammer_and_wrench: :magic_wand: :tada:

This is expected to ship in VS 2022 17.12 Preview 3.

**Pull request merged**